extensions/modules/file/src/test/resources/util/helper.xqm

```diff
@@ -0,0 +1,192 @@
+        (: cannot use fixtures here because this will lead to consumed input streams! :)
+        file:serialize-binary(
+            util:string-to-binary("SERVER_SECRET=123!"),
+            $directory || "/.env"),
+        file:serialize-binary(
+            util:string-to-binary("..."),
+            $directory || "/test/three.s")
+    )
+
+    return $directory
+};
+
+declare function helper:get-deleted-from-sync-result ($result as element(file:sync)) as xs:string* {
+    $result//file:delete/@name/string()
+};
+
+declare function helper:get-dir-from-sync-result ($result as element(file:sync)) as xs:string* {
+    $result/@file:dir/string()
+};
+
+declare function helper:get-updated-from-sync-result ($result as element(file:sync)) as xs:string* {
+    $result//file:update/@name/string()
+};
+
+declare function helper:list-files-and-directories ($directory as xs:string) as xs:string* {
+    file:list($directory)//(file:file|file:directory)/@name/string()
+};
+
+declare function helper:sync-with-options ($directory as xs:string, $options as item()?) as element(file:sync) {
+    file:sync($fixtures:collection, $directory, $options)/*
+};
+
+declare function helper:assert-sync-result (
+    $result as element(file:sync),
+    $expected as map(xs:string, xs:string*)
+) as xs:boolean {
+    helper:assert-permutation-of(
+        $expected?updated,
+        helper:get-updated-from-sync-result($result)
+    )
+    and
+    helper:assert-permutation-of(
+        $expected?deleted,
+        helper:get-deleted-from-sync-result($result)
+    )
+    and
+    helper:assert-permutation-of(
+        $expected?fs,
+        helper:get-dir-from-sync-result($result)
+        => helper:list-files-and-directories()
+    )
+};
+
+declare function helper:assert-permutation-of($expected as xs:anyAtomicType*, $actual as xs:anyAtomicType*) {
+    let $test := fold-left(
+        $expected,
+        [true(), $actual],
+        helper:permutation-reducer#2
+    )
+
+    return
+        if (not($test?1 or exists($test?2)))
+        then error($helper:error,
+        "Assertion failed: expected permutation of " ||
+               "(" || string-join($expected, ", ") || ")" ||
+               " but got (" || string-join($actual, ", ") || ")")
+        else true()
+};
+
+declare function helper:permutation-reducer ($result, $next) {
+    let $first-index := index-of($result?2, $next)[1]
+    return [
+        $result?1 and $first-index > 0,
+        helper:maybe-remove-item-at-index($result?2, $first-index)
+    ]
+};
+
+declare function helper:maybe-remove-item-at-index($sequence as xs:anyAtomicType*, $index as xs:integer?) {
+    if ($index = 1)
+    then subsequence($sequence, 2)
+    else if ($index > 1)
+    then (
+        subsequence($sequence, 1, $index - 1),
+        subsequence($sequence, $index + 1)
+    )
+    else $sequence (: do nothing - will be handled later :)
+};
+
+declare function helper:to-empty-sequence($_ as item()*) as empty-sequence() {};
```

Comment:
Remove this function - test code should be simple and dumb to aid in debugging. Introducing unnecessary complexity into test code is never a good idea.